drivers/net/nsos_adapt.c

```diff
@@ -0,0 +1,673 @@
+		wrap->addrinfo_mid.ai_addrlen = sizeof(wrap->addr_storage);
+
+		sockaddr_to_nsos_mid(res_p->ai_addr, res_p->ai_addrlen,
+				     wrap->addrinfo_mid.ai_addr, &wrap->addrinfo_mid.ai_addrlen);
+
+		wrap->addrinfo_mid.ai_canonname =
+			res_p->ai_canonname ? strdup(res_p->ai_canonname) : NULL;
+		wrap->addrinfo_mid.ai_next = &wrap[1].addrinfo_mid;
+	}
+
+	nsos_res_wraps[n_res - 1].addrinfo_mid.ai_next = NULL;
+
+	*mid_res = &nsos_res_wraps->addrinfo_mid;
+
+	return 0;
+
+free_wraps:
+	free(nsos_res_wraps);
+
+	return ret;
+}
+
+int nsos_adapt_getaddrinfo(const char *node, const char *service,
+			   const struct nsos_mid_addrinfo *hints_mid,
+			   struct nsos_mid_addrinfo **res_mid,
+			   int *system_errno)
+{
+	struct addrinfo hints;
+	struct addrinfo *res = NULL;
+	int ret;
+
+	if (hints_mid) {
+		hints.ai_flags = hints_mid->ai_flags;
+
+		ret = socket_family_from_nsos_mid(hints_mid->ai_family, &hints.ai_family);
+		if (ret < 0) {
+			*system_errno = -ret;
+			return NSOS_MID_EAI_SYSTEM;
+		}
+
+		ret = socket_type_from_nsos_mid(hints_mid->ai_socktype, &hints.ai_socktype);
+		if (ret < 0) {
+			*system_errno = -ret;
+			return NSOS_MID_EAI_SYSTEM;
+		}
+
+		ret = socket_proto_from_nsos_mid(hints_mid->ai_protocol, &hints.ai_protocol);
+		if (ret < 0) {
+			*system_errno = -ret;
+			return NSOS_MID_EAI_SYSTEM;
+		}
+	}
+
+	ret = getaddrinfo(node, service,
+			  hints_mid ? &hints : NULL,
+			  &res);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = addrinfo_to_nsos_mid(res, res_mid);
+	if (ret < 0) {
+		*system_errno = -ret;
+		return NSOS_MID_EAI_SYSTEM;
+	}
+
+	return ret;
+}
+
+void nsos_adapt_freeaddrinfo(struct nsos_mid_addrinfo *res_mid)
+{
+	struct nsos_addrinfo_wrap *wrap =
+		CONTAINER_OF(res_mid, struct nsos_addrinfo_wrap, addrinfo_mid);
+
+	freeaddrinfo(wrap->addrinfo);
+	free(wrap);
+}
+
+int nsos_adapt_fcntl_getfl(int fd)
+{
+	int flags;
+
+	flags = fcntl(fd, F_GETFL);
+
+	return fl_to_nsos_mid(flags);
+}
+
+int nsos_adapt_fcntl_setfl(int fd, int flags)
+{
+	int ret;
+
+	ret = fcntl(fd, F_SETFL, fl_from_nsos_mid(flags));
+	if (ret < 0) {
+		return -errno;
+	}
+
+	return 0;
+}
+
+static void nsos_adapt_poll_triggered(void)
+{
+	static struct pollfd *nsos_adapt_fds;
+	struct nsos_mid_pollfd *pollfd_mid;
+	struct pollfd *fds;
+	int nfds = sys_dlist_len(&nsos_adapt_sockets);
+	int ret;
+
+	fds = realloc(nsos_adapt_fds, sizeof(*fds) * nfds);
+	if (nfds > 0 && !fds) {
+		nsi_print_warning("error in nsos_adapt realloc()\n");
+		return;
+	}
+
+	nsos_adapt_fds = fds;
+
+	if (nfds == 0) {
+		nsos_adapt_poll_time = NSI_NEVER;
+		nsi_hws_find_next_event();
+		return;
+	}
+
+	nfds = 0;
+	SYS_DLIST_FOR_EACH_CONTAINER(&nsos_adapt_sockets, pollfd_mid, host_node) {
+		fds[nfds].fd = pollfd_mid->fd;
+		fds[nfds].events = pollfd_mid->events;
+		fds[nfds].revents = pollfd_mid->revents;
+		nfds++;
+	}
+
+	ret = poll(fds, nfds, 0);
+	if (ret < 0) {
+		nsi_print_warning("error in nsos_adapt poll(): errno=%d\n", errno);
```

Comment:
Question: should this be an error? (given that you disable the functionality after this)

Reply:
Extended poll() error handling with two cases:
- `EINTR` seems to be "recoverable", so left a warning message and changed `nsos_adapt_poll_time` to be `nsi_hws_get_time() + 1`
- other error codes: calling `nsi_print_error_and_exit()` instead, since further application run is most likely not reliable/useful